cache_utils/decorators.py

```diff
@@ -11,7 +11,8 @@
 logger = logging.getLogger("cache_utils")
 
 
-def cached(timeout, group=None, backend=None, key=None):
+def cached(timeout, group=None, backend=None, key=None,
+           filter_args_kwargs=None):
```

Comment:
This is a very useful feature :100: I have some opinions about the API for this, let me know your thoughts!

The goal here is to allow the user to modify the cache key for each call, right? `filter_args_kwargs` as a concept seems slightly too restrictive than it needs to be: we may also want to transform them. (Or if you want to live dangerously, add something new to the key, like a mutable global variable :feelsgood:)

Further, the return value doesn't have to be in the args & kwargs, tuple and dict form—we just need a serializable value we can use as a cache key. I think the simplest interface would be a function that maps arguments to any Python value that may be serialized to a concrete cache key.

I would separate out the following things and allow the user to customize each:

- Calculating the function cache key (the prefix)
- Based on function arguments on a single call, determine the unserialized cache key value
- Serializing cache key value above (JSON? Pickle? `hash`? `django-cache-utils` casts to string)

**Calculating the function cache key**

The generated function key in `django-cache-utils` is something like `cacheable_scrape_article[cached]:431`. We may want to make the `key` kwarg accept a callable to facilitate this, e.g.

```python
@cached(..., key=lambda fn: f'{fn.__name__}_version_1')
```

That would allow the default value of `key` to be the function that returns `cacheable_scrape_article[cached]:431` based on the function.

**Determining the cache key for a single call**

Specifying the cache key for a call would be facilitated with a keyword argument similar to `filter_args_kwargs`; I think the most intuitive kwarg name would be `key`, but that is already used for the function key. Without having to think about backwards compatibility, I would rename `key` to `fn_key` and have the call cache key function specified in `key`.

And for `key`, I would also consider calling the key function as `*args, **kwargs` of the initial call, versus the `args, kwargs` as used in `filter_args_kwargs`. The former allows binding names in the function signature for a terser, more explicit key function. Consider the following function we want to cache:

```python
def fetch_some_resource(url, timeout=None):
    return requests.get(url, timeout=timeout)
```

We want to cache this function, but only use `url` to fetch a cached value and ignore `timeout`. The simplest and easiest to read way to do that is like this:

```python
def key(url, *args, **kwargs):
    return url

@cached(3600, key=key)
def fetch_some_resource(url, timeout=None):
    return requests.get(url, timeout=timeout)
```

Whereas with a `filter_args_kwargs` requires a more verbose implementation if you want to surface the `url` binding:

```python
def filter_args_kwargs(args, kwargs):
    url, *rest = args
    return (url, ), {}
```

**Serializing the key value for a call**

Now that we have the key (Python) value for a call, we can use the user-specified serializer to turn it into a key value usable by the cache backend. `django-cache-utils` uses [a simple string cast](https://github.com/UpbeatPR/django-cache-utils/blob/master/cache_utils/utils.py#L27), but I think JSON/pickle would be more robust. Then the whole cache key would be `serialize(fn_key(fn)) + serialize(call_key(*args, **kwargs))`

-----------------------------------------

Obviously we don't need to make all these changes at once, but it would be useful to have a consistent, robust API in mind as we make improvements.

Reply:
I agree that renaming `key` to `fn_key` and making it a callable with the default work as is now would be nice.

Also now that `key` is free, we can use that for computing the paramters to a python value, I wanted to use that but was taken.

Lastly, I guess a nice to have would be to add other serialization methods, yep. But this isn't an immediate option to use since we use cached in a bunch of places and changing this to sth other than the `_args_to_unicode` with prefix `[cached]` would be equivalent to invalidating all our cache. We could start using different serialization protocols for new code though.

I've changed the code to support this. At first I was hesitant to do too many changes and the odd implementation from before encouraged hacks. I like your idea with splitting the args and the function conversion to serializables and using those. I just used str for now, I didn't look much into what the final output needs to be so memcached/group_backend.py etc. related code would expect. Moreover, kept backwards compatibility so our old cache is hit with this update. Let me know what you think and if you spot any issues 👍 .